### PR TITLE
Wct/mvnormal mean constraints

### PR DESCRIPTION
**Pull request by matbesancon**

### CODE REVIEW

File: src/multivariate/mvlognormal.jl [hunk 1]
```diff
@@ -161,18 +161,18 @@ Mean vector ``\\boldsymbol{\\mu}`` and covariance matrix ``\\boldsymbol{\\Sigma}
 underlying normal distribution are known as the *location* and *scale*
 parameters of the corresponding lognormal distribution.
 """
-struct MvLogNormal{T<:Real,Cov<:AbstractPDMat,Mean<:Union{Vector, ZeroVector}} <: AbstractMvLogNormal
+struct MvLogNormal{T<:Real,Cov<:AbstractPDMat,Mean<:Union{AbstractVector, ZeroVector}} <: AbstractMvLogNormal
```
Comment (mschauer) [inline src/multivariate/mvlognormal.jl]: ZeroVector is an abstract vector

Comment (matbesancon) [reply src/multivariate/mvlognormal.jl]: good point, I'll simplify this

File: src/multivariate/mvnormal.jl [hunk 2]
```diff
@@ -79,10 +79,10 @@ modes(d::AbstractMvNormal) = [mean(d)]
 
 function entropy(d::AbstractMvNormal)
     ldcd = logdetcov(d)
-    (length(d) * (typeof(ldcd)(log2π) + 1) + ldcd)/2
+    (length(d) * (typeof(ldcd)(log2π) + 1.) + ldcd)/2.
```
Comment (mschauer) [inline src/multivariate/mvnormal.jl]: Wasn't this fine?

File: src/multivariate/mvnormal.jl [hunk 3]
```diff
@@ -162,7 +162,7 @@ isotropic covariance as `abs2(sig) * eye(d)`.
 **Note:** The constructor will choose an appropriate covariance form internally, so that
 special structure of the covariance can be exploited.
 """
-struct MvNormal{T<:Real,Cov<:AbstractPDMat,Mean<:Union{AbstractVector, ZeroVector}} <: AbstractMvNormal
+struct MvNormal{T<:Real,Cov<:AbstractPDMat,Mean<:AbstractVector}} <: AbstractMvNormal
```
Comment (mschauer) [inline src/multivariate/mvnormal.jl]: The double brace causes the error.

File: src/multivariate/mvnormal.jl (hunk 2 above)
Comment (matbesancon) [reply src/multivariate/mvnormal.jl]: we can avoid necessary type promotions, these happen at runtime?

Comment (matbesancon) [reply src/multivariate/mvnormal.jl]: but point, this is forcing everything to Float64

**Comment by matbesancon** [comment]
Fixes #831

File: src/multivariate/mvlognormal.jl [hunk 4]
```diff
@@ -161,18 +161,18 @@ Mean vector ``\\boldsymbol{\\mu}`` and covariance matrix ``\\boldsymbol{\\Sigma}
...
 parameters of the corresponding lognormal distribution.
 """
-struct MvLogNormal{T<:Real,Cov<:AbstractPDMat,Mean<:Union{Vector, ZeroVector}} <: AbstractMvLogNormal
+struct MvLogNormal{T<:Real,Cov<:AbstractPDMat,Mean<:AbstractVector} <: AbstractMvLogNormal
     normal::MvNormal{T,Cov,Mean}
 end
 
 #Constructors mirror the ones for MvNormmal
-MvLogNormal(μ::Union{Vector,ZeroVector},Σ::AbstractPDMat) = MvLogNormal(MvNormal(μ,Σ))
+MvLogNormal(μ::M,Σ::AbstractPDMat) where {M<:AbstractVector} = MvLogNormal(MvNormal(μ,Σ))
```
Comment (ararslan) [inline src/multivariate/mvlognormal.jl]:
```suggestion
MvLogNormal(μ::AbstractVector,Σ::AbstractPDMat) = MvLogNormal(MvNormal(μ,Σ))
```

File: test/mvlognormal.jl [hunk 5]
```diff
@@ -92,46 +92,44 @@ end
...
         (MvLogNormal(3, sqrt(0.2)), zeros(3), Matrix(0.2I, 3, 3)),
-        (MvLogNormal(mu, sqrt.(va)), mu, Matrix(Diagonal(va))),
-        (MvLogNormal(sqrt.(va)), zeros(3), Matrix(Diagonal(va))),
+        (MvLogNormal(mu, Vector{Float64}(sqrt.(va))), mu, Matrix(Diagonal(va))), # Julia 0.4 loses type information so Vector{Float64} can be dropped when we don't support 0.4
+        (MvLogNormal(Vector{Float64}(sqrt.(va))), zeros(3), Matrix(Diagonal(va))), # Julia 0.4 loses type information so Vector{Float64} can be dropped when we don't support 0.4
         (MvLogNormal(mu, C), mu, C),
         (MvLogNormal(C), zeros(3), C) ]
-        
+
+        println("    testing $(typeof(g)) with normal distribution $(Distributions.distrname(g.normal))")
```
Comment (ararslan) [inline test/mvlognormal.jl]: I think this could just be another `@testset`.

Comment (matbesancon) [reply test/mvlognormal.jl]: splitting which part? The only test in the loop are the three below this line

File: src/multivariate/mvnormal.jl [hunk 6]
```diff
@@ -178,36 +179,35 @@ const ZeroMeanDiagNormal = MvNormal{Float64,PDiagMat{Float64,Vector{Float64}},Ze
...
-MvNormal(μ::Vector{T}, σ::Vector{T}) where {T<:Real} = MvNormal(μ, PDiagMat(abs2.(σ)))
-MvNormal(μ::Vector{T}, σ::T) where {T<:Real} = MvNormal(μ, ScalMat(length(μ), abs2(σ)))
+MvNormal(μ::AbstractVector{T}, Σ::Matrix{T}) where {T<:Real} = MvNormal(μ, PDMat(Σ))
+MvNormal(μ::AbstractVector{T}, Σ::Union{Symmetric{T}, Hermitian{T}}) where {T<:Real} = MvNormal(μ, PDMat(Σ))
+MvNormal(μ::AbstractVector{T}, Σ::Diagonal{T}) where {T<:Real} = MvNormal(μ, PDiagMat(diag(Σ)))
+MvNormal(μ::AbstractVector{T}, σ::Vector{T}) where {T<:Real} = MvNormal(μ, PDiagMat(abs2.(σ)))
+MvNormal(μ::AbstractVector{T}, σ::T) where {T<:Real} = MvNormal(μ, ScalMat(length(μ), abs2(σ)))
 
-function MvNormal(μ::Vector{T}, Σ::VecOrMat{S}) where {T<:Real,S<:Real}
+function MvNormal(μ::AbstractVector{T}, Σ::VecOrMat{S}) where {T<:Real,S<:Real}
```
Comment (willtebbutt) [inline src/multivariate/mvnormal.jl]: Style (please feel free to ignore): It looks to me that `T` and `S` aren't used inside the function at all, so could we make the signature:
```julia
MvNormal(μ::AbstractVector{<:Real}, Σ::VecOrMat{<:Real})
```
instead? (The same applies to a couple of other methods below)

File: src/multivariate/mvnormal.jl [hunk 7]
```diff
@@ -256,7 +256,7 @@ sqmahal(d::MvNormal, x::AbstractVector) = invquad(d.Σ, broadcast(-, x, d.μ))
 sqmahal!(r::AbstractVector, d::MvNormal, x::AbstractMatrix) =
     invquad!(r, d.Σ, broadcast(-, x, d.μ))
 
-gradlogpdf(d::MvNormal, x::Vector) = -(d.Σ \ broadcast(-, x, d.μ))
+gradlogpdf(d::MvNormal, x::AbstractVector) = -(d.Σ \ broadcast(-, x, d.μ))
```
Comment (willtebbutt) [inline src/multivariate/mvnormal.jl]: Should `x` be `x::AbstractVector{<:Real}`?

Also, style, `broadcast(-, x, d.μ)` -> `(x .- d.μ)`?

File: src/multivariate/mvnormalcanon.jl [hunk 8]
```diff
@@ -63,7 +63,7 @@ Construct a multivariate normal distribution of dimension `d`, with zero mean an
 
 **Note:** `MvNormalCanon` share the same set of methods as `MvNormal`.
 """
-struct MvNormalCanon{T<:Real,P<:AbstractPDMat,V<:Union{Vector,ZeroVector}} <: AbstractMvNormal
+struct MvNormalCanon{T<:Real,P<:AbstractPDMat,V<:Union{AbstractVector,ZeroVector}} <: AbstractMvNormal
```
Comment (willtebbutt) [inline src/multivariate/mvnormalcanon.jl]: `Union{AbstractVector,ZeroVector}}` -> `AbstractVector`?

File: src/multivariate/mvtdist.jl [hunk 9]
```diff
@@ -39,7 +39,7 @@ function convert(::Type{GenericMvTDist{T}}, d::GenericMvTDist) where T<:Real
     S = convert(AbstractArray{T}, d.Σ)
     GenericMvTDist{T, typeof(S)}(T(d.df), d.dim, d.zeromean, convert(AbstractArray{T}, d.μ), S)
 end
-function convert(::Type{GenericMvTDist{T}}, df, dim, zeromean, μ::Union{Vector, ZeroVector}, Σ::AbstractPDMat) where T<:Real
+function convert(::Type{GenericMvTDist{T}}, df, dim, zeromean, μ::Union{AbstractVector, ZeroVector}, Σ::AbstractPDMat) where T<:Real
```
Comment (willtebbutt) [inline src/multivariate/mvtdist.jl]: `Union{AbstractVector, ZeroVector}` -> `AbstractVector`?

File: src/multivariate/mvnormal.jl [hunk 10]
```diff
@@ -256,7 +256,7 @@ sqmahal(d::MvNormal, x::AbstractVector) = invquad(d.Σ, broadcast(-, x, d.μ))
 sqmahal!(r::AbstractVector, d::MvNormal, x::AbstractMatrix) =
     invquad!(r, d.Σ, broadcast(-, x, d.μ))
 
-gradlogpdf(d::MvNormal, x::AbstractVector) = -(d.Σ \ broadcast(-, x, d.μ))
+gradlogpdf(d::MvNormal, x::AbstractVector{<:Real}) = -(d.Σ \ x .- d.μ)
```
Comment (willtebbutt) [inline src/multivariate/mvnormal.jl]: I suspect that this needs bracketing. i.e. `-(d.Σ \ (x .- d.μ))`

Comment (matbesancon) [reply src/multivariate/mvnormal.jl]: fixed :)

File: src/multivariate/mvnormal.jl [hunk 11]
```diff
@@ -178,36 +179,35 @@ const ZeroMeanDiagNormal = MvNormal{Float64,PDiagMat{Float64,Vector{Float64}},Ze
...
+function MvNormal(Σ::Cov) where {T, Cov<:AbstractPDMat{T}}
     MvNormal{T,Cov,ZeroVector{T}}(ZeroVector(T, dim(Σ)), Σ)
 end
 
-MvNormal(μ::Vector{T}, Σ::Matrix{T}) where {T<:Real} = MvNormal(μ, PDMat(Σ))
-MvNormal(μ::Vector{T}, Σ::Union{Symmetric{T}, Hermitian{T}}) where {T<:Real} = MvNormal(μ, PDMat(Σ))
-MvNormal(μ::Vector{T}, Σ::Diagonal{T}) where {T<:Real} = MvNormal(μ, PDiagMat(diag(Σ)))
-MvNormal(μ::Vector{T}, σ::Vector{T}) where {T<:Real} = MvNormal(μ, PDiagMat(abs2.(σ)))
-MvNormal(μ::Vector{T}, σ::T) where {T<:Real} = MvNormal(μ, ScalMat(length(μ), abs2(σ)))
+MvNormal(μ::AbstractVector{T}, Σ::Matrix{T}) where {T<:Real} = MvNormal(μ, PDMat(Σ))
```
[the remark this answers is not in the document]
Comment (matbesancon) [reply src/multivariate/mvnormal.jl]: It seems reasonable to me yes

File: src/multivariate/mvnormalcanon.jl [hunk 12]
```diff
@@ -100,28 +104,29 @@ function MvNormalCanon(J::P) where P<:AbstractPDMat
     MvNormalCanon{eltype(J),P,ZeroVector{eltype(J)}}(z, z, J)
 end
 
-function MvNormalCanon(h::Vector{T}, J::P) where {T<:Real, P<:AbstractPDMat}
+function MvNormalCanon(h::AbstractVector{T}, J::P) where {T<:Real, P<:AbstractPDMat}
     length(h) == dim(J) || throw(DimensionMismatch("Inconsistent argument dimensions"))
     R = Base.promote_eltype(h, J)
-    hh, JJ = convert(AbstractArray{R}, h), convert(AbstractArray{R}, J)
+    hh, JJ = collect(convert(AbstractArray{R}, h)), convert(AbstractArray{R}, J)
```
Comment (mschauer) [inline src/multivariate/mvnormalcanon.jl]: Was this collect needed somewhere?

File: src/multivariate/mvnormalcanon.jl [hunk 13]
```diff
@@ -149,7 +154,10 @@ meanform(d::MvNormalCanon) = MvNormal(d.μ, inv(d.J))
 # meanform{C, T<:Real}(d::MvNormalCanon{T,C,Vector{T}}) = MvNormal(d.μ, inv(d.J))
 # meanform{C, T<:Real}(d::MvNormalCanon{T,C,ZeroVector{T}}) = MvNormal(inv(d.J))
 
-canonform(d::MvNormal{T,C,Vector{T}}) where {C, T<:Real} = (J = inv(d.Σ); MvNormalCanon(d.μ, J * d.μ, J))
+function canonform(d::MvNormal{T,C,<:AbstractVector{T}}) where {C, T<:Real}
+    J = inv(d.Σ)
+    return MvNormalCanon(d.μ, J * collect(d.μ), J)
```
Comment (mschauer) [inline src/multivariate/mvnormalcanon.jl]: Same here?

File: src/multivariate/mvnormal.jl [hunk 14]
```diff
@@ -256,7 +256,7 @@ sqmahal(d::MvNormal, x::AbstractVector) = invquad(d.Σ, broadcast(-, x, d.μ))
 sqmahal!(r::AbstractVector, d::MvNormal, x::AbstractMatrix) =
     invquad!(r, d.Σ, broadcast(-, x, d.μ))
 
-gradlogpdf(d::MvNormal, x::Vector) = -(d.Σ \ broadcast(-, x, d.μ))
+gradlogpdf(d::MvNormal, x::AbstractVector{<:Real}) = -(d.Σ \ (x .- d.μ))
```
[the remark this answers is not in the document]
Comment (matbesancon) [reply src/multivariate/mvnormal.jl]: > You might want to do automatic differentiation with respect to the mean vector in a likelihood...

You mean for the same type constraint? I see yes, hadn't thought of this point

File: src/multivariate/mvnormal.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (jmxpearson) [reply src/multivariate/mvnormal.jl]: If the two types are not the same, will it cause difficulty with other code? Don't some functions implicitly assume that μ and Σ are parameterized on the same type?

File: src/multivariate/mvnormalcanon.jl (hunk 12 above)
Comment (jmxpearson) [reply src/multivariate/mvnormalcanon.jl]: In case `h` is `I` or `ZeroVector`?

Comment (matbesancon) [reply src/multivariate/mvnormalcanon.jl]: yes, some linear algebra operations are not defined on `AbstractArrays` I think

**Comment by matbesancon** [comment]
I will check the suggested bound relaxation tomorrow morning

File: src/multivariate/mvnormal.jl (hunk 14 above)
[the remark this answers is not in the document]
Comment (mschauer) [reply src/multivariate/mvnormal.jl]: Something in the order of comments went wrong.

[the remark this answers is not in the document]
Comment (matbesancon) [reply src/multivariate/mvnormal.jl]: oh ok I get it now, the order displayed was really odd

**Comment by matbesancon** [comment]
Fixed the broadcast thing, thanks @mschauer, merging as soon as green

**Pull request merged**